File: blind75/go/maximumproductsubarray.go
func maxProduct(nums []int) int {
    if len(nums) == 0 {
        return 0
    }

    maxProduct := nums[0]
    minProduct := nums[0]
    result := nums[0]

    for i := 1; i < len(nums); i++ {
        tempMax := maxProduct
        maxProduct = max(nums[i], max(nums[i]*maxProduct, nums[i]*minProduct))
        minProduct = min(nums[i], min(nums[i]*tempMax, nums[i]*minProduct))
        result = max(result, maxProduct)
    }
    return result
}

File: blind75/php/maximumproductsubarray.php
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function maxProduct($nums) {
        if (empty($nums)) return 0;

        $max_product = $nums[0];
        $min_product = $nums[0];
        $result = $nums[0];

        $length = count($nums);
        for ($i = 1; $i < $length; $i++) {
            $temp_max = $max_product;
            $max_product = max($nums[$i], $nums[$i] * $max_product, $nums[$i] * $min_product);
            $min_product = min($nums[$i], $nums[$i] * $temp_max, $nums[$i] * $min_product);
            $result = max($result, $max_product);
        }
        return $result;
    }
}
